platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/WorkNodeCodec.kt

```diff
@@ -55,81 +76,319 @@ typealias IdForNode = (Node) -> Int
 class WorkNodeCodec(
     private val owner: GradleInternal,
     private val internalTypesCodec: Codec<Any?>,
-    private val ordinalGroups: OrdinalGroupFactory
+    private val ordinalGroups: OrdinalGroupFactory,
+    private val contextSource: IsolateContextSource,
+    /** Should we store work nodes in parallel? */
+    private val parallelStore: Boolean,
+    /** Should we load work nodes in parallel? */
+    private val parallelLoad: Boolean
 ) {
 
-    suspend fun WriteContext.writeWork(work: ScheduledWork) {
+    fun WriteContext.writeWork(work: ScheduledWork) {
         // Share bean instances across all nodes (except tasks, which have their own isolate)
         withGradleIsolate(owner, internalTypesCodec) {
             doWrite(work)
         }
     }
 
-    suspend fun ReadContext.readWork(): ScheduledWork =
+    fun ReadContext.readWork(): ScheduledWork =
         withGradleIsolate(owner, internalTypesCodec) {
             doRead()
         }
 
     private
-    suspend fun WriteContext.doWrite(work: ScheduledWork) {
+    fun WriteContext.doWrite(work: ScheduledWork) {
         val nodes = work.scheduledNodes
+        val entryNodes = work.entryNodes
         val nodeCount = nodes.size
-        writeSmallInt(nodeCount)
-        val scheduledNodeIds = Object2IntOpenHashMap<Node>(nodeCount)
+        val scheduledNodeIds = Object2IntOpenHashMap<Node>(nodeCount).apply {
+            defaultReturnValue(-1)
+        }
+
+        val scheduledEntryNodeIds = assignNodeIds(scheduledNodeIds, nodes, entryNodes)
+        val actionNodePostExecutionSuccessors = writeNodes(nodes, scheduledNodeIds)
```

Comment:
💅 `writeNodes` should receive the `idForNode` function instead of the full collection to reinforce the idea that the collection is effectively immutable at this point and to get the invariant assertion.